[FLINK-3311/FLINK-3332] Add Cassandra connector

zentol opened the pull request:
This PR is a combination of #1640 and #1660, essentially providing a mid-merge view of both PR's. As they reside in the same module and expose similar functionalities, several things were changed in both commits to make the code more similar.

Such changes include formatting, field ordering, code in-lining, package declarations, file names, test unification, and the addition of a central method to create sinks (CassandraSink.addSink(DataStream<T> input)) to reduce complexity for the user resulting from different sink implementations.

alkagin commented:
Great work :+1:

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/pom.xml:
```diff
@@ -0,0 +1,175 @@
+		<dependency>
+			<groupId>com.google.guava</groupId>
+			<artifactId>guava</artifactId>
+			<version>16.0.1</version>
```
Please use the `guava.version` variable here to use the same guava version across Flink

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraInputFormat.java:
```diff
@@ -0,0 +1,129 @@
+		try {
+			cluster.close();
+		} catch (Exception e) {
+			LOG.info("Inputformat couldn't be closed - " + e.getMessage());
```
I would add the cause in both cases.

alkagin commented:
Hi, 
I am testing the Cassandra connector and I may have found a bug using CassandraSink-DSL. During the execution it throws `java.io.NotSerializableException: org.apache.flink.streaming.connectors.cassandra.CassandraSink$CassandraTupleSinkBuilder` CassandraSinkBuilder doesn't implement java.io.Serializable. I forked and added java.io.serializable to CassandraSinkBuilder, after that it throws `java.io.NotSerializableException: org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator`.
Not using the DSL it works as expected. 
Job source code: https://gist.github.com/alkagin/6620ebdd10ef6bb47eee

Hope it helps.

zentol commented:
@alkagin Thank you for reporting this, will look into it. I think i may already know how to fix it...

alkagin commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraSink.java:
```diff
@@ -0,0 +1,282 @@
+	}
+
+	public enum ConsistencyLevel {
+		At_LEAST_ONCE,
```
typo: AT_LEAST_ONCE

alkagin commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericAtLeastOnceSink.java:
```diff
@@ -0,0 +1,197 @@
+ *
+ * @param <IN> Type of the elements emitted by this sink
+ */
+public abstract class GenericAtLeastOnceSink<IN> extends AbstractStreamOperator<IN> implements OneInputStreamOperator<IN, IN> {
```
GenericExactlyOnceSink

zentol replied:
I've struggled a bit with the name for this class, but came to the conclusion that since class itself does not provide exactly-once guarantees for all use-cases. As such the GenericAtLeastOnceSink name is intended.

alkagin replied:
Ok, I misunderstood CassandraIdempotentExactlyOnceSink's inheritance.

zentol commented:
Just pushed the following changes:
* Fixed AT_LEAST_ONCE type
* CassandraInputFormat.close properly propagates exceptions in close()
* cassandra pom now uses guava.version attribute
* fixed NotSerializableException For CassandraAtLeastOnceTuple-/PojoSink

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraOutputFormat.java:
```diff
@@ -0,0 +1,123 @@
+		try {
+			session.close();
+		} catch (Exception e) {
+			LOG.info("Inputformat couldn't be closed - " + e.getMessage());
```
I would make this a WARN and forward the cause to the log entry

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraCommitter.java:
```diff
@@ -0,0 +1,126 @@
+	private transient Session session;
+
+	private static final String KEYSPACE = "flink_auxiliary";
+	private String TABLE = "checkpoints_";
```
by convention, (non static) fields are lowercase

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraCommitter.java:
```diff
@@ -0,0 +1,126 @@
+		session = cluster.connect();
+
+		session.execute(String.format("CREATE KEYSPACE IF NOT EXISTS %s with replication={'class':'SimpleStrategy', 'replication_factor':3};", KEYSPACE));
+		session.execute(String.format("CREATE TABLE IF NOT EXISTS %s.%s (sink_id text, sub_id int, checkpoint_id bigint, PRIMARY KEY (sink_id, sub_id));", KEYSPACE, TABLE));
```
I think it would be better to allow users passing a custom keyspace.

rmetzger replied:
Users still can not pass custom keyspaces

zentol replied:
fixed

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraSink.java:
```diff
@@ -0,0 +1,313 @@
+ * @param <IN> input type
+ */
+public class CassandraSink<IN> {
+	private static final String jobID = UUID.randomUUID().toString().replace("-", "_");
```
`jobID` is used within Flink a lot and has a different meaning there. Maybe it make sense to rename this to "sink id".

Why is this field static? This can lead to problems when a user is using two cassandra sinks

zentol replied:
why should it be named sinkID when it is supposed to be a job-specific ID?

rmetzger replied:
The problem is that there will be two `jobID`s which can lead to confusions.

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraSink.java:
```diff
@@ -0,0 +1,313 @@
+		 * @return this builder
+		 */
+		public CassandraSinkBuilder<IN> setHost(final String host, final int port) {
+			builder = new ClusterBuilder() {
```
I wonder whether we should throw an exception if the builder is already set. Otherwise, if somebody first sets a builder and then a host the results will be very misleading.

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraSink.java:
```diff
@@ -0,0 +1,313 @@
+		 */
+		public abstract CassandraSink<IN> build() throws Exception;
+		
+		protected void sanityCheck() {
```
very good!

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraSinkBase.java:
```diff
@@ -0,0 +1,93 @@
+
+			@Override
+			public void onFailure(Throwable t) {
+				exception = t;
```
Can you log the exception as well. Maybe invoke() is never called after the failure and nobody knows what's going on

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraTupleWriteAheadSink.java:
```diff
@@ -0,0 +1,136 @@
+
+			@Override
+			public void onFailure(Throwable throwable) {
+				exception = throwable;
```
Error logging here as well

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraTupleWriteAheadSink.java:
```diff
@@ -0,0 +1,136 @@
+		while (updatesSent != updatesConfirmed.get()) {
+			try {
+				Thread.sleep(100);
+			} catch (InterruptedException e) {
```
I would leave the while loop on an interrupt.
It might happen that the thread gets locked in the loop because it can not be interrupted.

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/ClusterBuilder.java:
```diff
@@ -0,0 +1,31 @@
+
+import java.io.Serializable;
+
+public abstract class ClusterBuilder implements Serializable {
```
Since this is a user facing class, I would add some javadocs

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/example/BatchExample.java:
```diff
@@ -0,0 +1,71 @@
+
+import java.util.ArrayList;
+
+public class BatchExample {
```
Javadocs

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/example/BatchExample.java:
```diff
@@ -0,0 +1,71 @@
+
+	/*
+	 *	table script: "CREATE TABLE test.batches (number int, strings text, PRIMARY KEY(number, strings));"
+	 */
```
I would remove this comment and move the query to the class javadocs.

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/example/CassandraTupleWriteAheadSinkExample.java:
```diff
@@ -0,0 +1,88 @@
+public class CassandraTupleWriteAheadSinkExample {
+	public static void main(String[] args) throws Exception {
+
+		class MySource implements SourceFunction<Tuple2<String, Integer>>, Checkpointed<Integer> {
```
I would move the class out of the main method

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/test/resources/cassandra.yaml:
```diff
@@ -0,0 +1,43 @@
+commitlog_directory: $PATH\commit'
+data_file_directories:
+    - $PATH\data'
+saved_caches_directory: $PATH\cache'
```
I wonder why the path works on unix platforms. Only windows is using `\` as a separator.

zentol replied:
the simple explanation would be that the EmbeddedCassandraService can resolve the path properly regardless of platform.

zentol commented:
@rmetzger addressed all your comments.

rmetzger commented:
I've tested the change again on a cluster and locally. The mode without WAL works fine, the WAL-variant fails on (or quickly after) recovery.
It seems to be an issue with the `FsStatebackend`, not with this PR. Still I'll wait with the merging until we understood the issue.

![image](https://cloud.githubusercontent.com/assets/89049/14383013/6ed277de-fd92-11e5-8387-4e09575b9cce.png)

![image](https://cloud.githubusercontent.com/assets/89049/14383018/776e0368-fd92-11e5-9bfe-cce5fd9d0277.png)

zentol commented:
@rmetzger I hav efixed the issue you encountered. The problem was that the JM tries to discard all state belonging to pending checkpoints upon job failure, which effectively purged the write-ahead log. The simple fix is to not do anything in ExactlyOnceState#discardState.

rmetzger commented:
Thank you for the fix. I'll take a look at it soon.

alkagin commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraSink.java:
```diff
@@ -0,0 +1,316 @@
+		 * @return this builder
+		 */
+		public CassandraSinkBuilder<IN> setClusterBuilder(ClusterBuilder builder) {
+			if (builder != null) {
```
`this.builder`instead of `builder`, correct?

zentol replied:
yes. i distinctly recall having fixed this...

alkagin commented:
Hi guys :) any update on this?

rmetzger commented:
I'll test the PR again

rmetzger commented:
I just tried the PR, but the recovery after a failure doesn't seem to work:

```
java.lang.RuntimeException: Error triggering a checkpoint as the result of receiving checkpoint barrier
	at org.apache.flink.streaming.runtime.tasks.StreamTask$2.onEvent(StreamTask.java:681)
	at org.apache.flink.streaming.runtime.tasks.StreamTask$2.onEvent(StreamTask.java:674)
	at org.apache.flink.streaming.runtime.io.BarrierBuffer.processBarrier(BarrierBuffer.java:203)
	at org.apache.flink.streaming.runtime.io.BarrierBuffer.getNextNonBlocked(BarrierBuffer.java:129)
	at org.apache.flink.streaming.runtime.io.StreamInputProcessor.processInput(StreamInputProcessor.java:175)
	at org.apache.flink.streaming.runtime.tasks.OneInputStreamTask.run(OneInputStreamTask.java:65)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:224)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:559)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: Failed to fetch state handle size
	at org.apache.flink.runtime.taskmanager.RuntimeEnvironment.acknowledgeCheckpoint(RuntimeEnvironment.java:234)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.triggerCheckpoint(StreamTask.java:511)
	at org.apache.flink.streaming.runtime.tasks.StreamTask$2.onEvent(StreamTask.java:678)
	... 8 more
Caused by: java.io.FileNotFoundException: File does not exist: hdfs://nameservice1/user/robert/cassandra-fs/e70d0b78b7875877f42a8ebfba463f14/chk-0/9f892bc0-b5e2-484f-a981-6e666e7ad897
	at org.apache.hadoop.hdfs.DistributedFileSystem$18.doCall(DistributedFileSystem.java:1122)
	at org.apache.hadoop.hdfs.DistributedFileSystem$18.doCall(DistributedFileSystem.java:1114)
	at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81)
	at org.apache.hadoop.hdfs.DistributedFileSystem.getFileStatus(DistributedFileSystem.java:1114)
	at org.apache.flink.runtime.fs.hdfs.HadoopFileSystem.getFileStatus(HadoopFileSystem.java:351)
	at org.apache.flink.runtime.state.filesystem.AbstractFileStateHandle.getFileSize(AbstractFileStateHandle.java:93)
	at org.apache.flink.runtime.state.filesystem.FileStreamStateHandle.getStateSize(FileStreamStateHandle.java:58)
	at org.apache.flink.runtime.state.AbstractStateBackend$DataInputViewHandle.getStateSize(AbstractStateBackend.java:428)
	at org.apache.flink.streaming.runtime.operators.GenericAtLeastOnceSink$ExactlyOnceState.getStateSize(GenericAtLeastOnceSink.java:190)
	at org.apache.flink.streaming.runtime.tasks.StreamTaskStateList.getStateSize(StreamTaskStateList.java:81)
	at org.apache.flink.runtime.taskmanager.RuntimeEnvironment.acknowledgeCheckpoint(RuntimeEnvironment.java:231)
	... 10 more
```

tillrohrmann commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericAtLeastOnceSink.java:
```diff
@@ -0,0 +1,192 @@
+	 * This state is used to keep a list of all StateHandles (essentially references to past OperatorStates) that were
+	 * used since the last completed checkpoint.
+	 **/
+	public class ExactlyOnceState implements StateHandle<Serializable> {
```
Serial version uid missing

tillrohrmann commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericAtLeastOnceSink.java:
```diff
@@ -0,0 +1,192 @@
+			Set<Long> checkpointsToRemove = new HashSet<>();
+			for (Long pastCheckpointId : pastCheckpointIds) {
+				if (pastCheckpointId <= checkpointId) {
+					if (!committer.isCheckpointCommitted(pastCheckpointId)) {
```
Does this always trigger a lookup? Wouldn't it be more efficient to fetch the last committed checkpoint id and then use this value for the comparison?

zentol replied:
whether it triggers a lookup is up to the Committer implementation. You could always keep a local version, and only lookup the value when the local version is null.

tillrohrmann replied:
Maybe we should do this for the `CassandraCommitter` implementation.

zentol replied:
I've implemented this.

tillrohrmann commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericAtLeastOnceSink.java:
```diff
@@ -0,0 +1,192 @@
+						Tuple2<Long, StateHandle<DataInputView>> handle = state.pendingHandles.get(pastCheckpointId);
+						DataInputView in = handle.f1.getState(getUserCodeClassloader());
+						sendValues(new ReusingMutableToRegularIteratorWrapper<>(new InputViewIterator<>(in, serializer), serializer), handle.f0);
+						committer.commitCheckpoint(pastCheckpointId);
```
Can it be that committing the checkpoint id can fail (with cassandra)? Is the commit checkpoint operation synchronous (CassandraCommitter)?

zentol replied:
yes it is synchronous. should it fail an exception is thrown.

tillrohrmann commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraCommitter.java:
```diff
@@ -0,0 +1,131 @@
+
+	@Override
+	public void commitCheckpoint(long checkpointID) {
+		session.execute(updateStatement.bind(checkpointID));
```
What happens if the `execute` call fails? Can we check the `ResultSet` for the success of the operation? If the commit did not succeed, then I think we should throw an exception.

zentol replied:
an exception is thrown in case of failure.

tillrohrmann replied:
Ah ok, I see. `Session` throws` an unchecked exception.

tillrohrmann commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraCommitter.java:
```diff
@@ -0,0 +1,131 @@
+		try {
+			session.close();
+		} catch (Exception e) {
+			LOG.error("Error while closing session.", e);
```
Why not propagating the exception here and simply logging it? Shouldn't the whole thing fail if we cannot create the necessary resources?

zentol replied:
session.execute() throws an exception if the resource cannot be created. This exception is propagated, which is the important part.

exceptions while closing the session isn't particularly important.

tillrohrmann replied:
This means that a failed closing operation cannot leave the external system in a corrupted state?

zentol replied:
correct.

tillrohrmann commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraCommitter.java:
```diff
@@ -0,0 +1,131 @@
+		try {
+			session.close();
+		} catch (Exception e) {
+			LOG.error("Error while closing session.", e);
```
Why swallowing the exception here? Shouldn't it be propagated to the executing task?

zentol replied:
because some random exception caused by session.close() (like an NPE) shouldn't cause other exceptions to be hidden.

zentol commented:
I finally was able to fix the restart issue. there were 2 massive bugs in the CassandraCommitter:
- within open() the checkpoint entry was always overridden
- within close() the checkpoint entry was always deleted

in addition i have made the following changes:
- renamed GenericAtLeastOnceSink to GenericWriteAheadSink
- implemented a caching of the last committed checkpointID in the CassandraCommitter

It's rather obvious that more tests are required.

rmetzger commented:
I've reviewed the connector again.
The issues I've seen previously (failure on restart) are resolved.
However, I found new issues:
- The Cassandra Sink doesn't fail (at least not within 15 minutes) if Cassandra is not available anymore. Its probably just a configuration setting of the cassandra driver to fail after a certain amount of time.
- We should probably introduce a (configurable) limit (nr. records / some gb's) for the write ahead log. It seemed to me, that due to the failed other instance, no checkpoints were able to complete anymore (because some of the cassandra sinks were stuck in the notifyCheckpointComplete()), while other's were accepting data into the WAL. This lead to a lot of data being written into the statebackend. I think the cassandra sink should stop at some point in such a situation.

Also, I would like to test the exactly once behavior on a cluster more thoroughly. Currently, I've only tested whether the connector is properly failing and restoring, but I didn't test if the written data is actually correct.

However, since the code seems to be working under normal operation, I would suggest to merge the connector now, and then file follow up JIRAs for the remaining issues.
This makes collaboration and reviews easier and allows our users to help testing the cassandra connector.



Some log:
```
2016-06-03 12:28:36,478 ERROR org.apache.flink.streaming.runtime.operators.GenericWriteAheadSink  - Error while sending value.
com.datastax.driver.core.exceptions.UnavailableException: Not enough replicas available for query at consistency LOCAL_ONE (1 required but only 0 alive)
	at com.datastax.driver.core.exceptions.UnavailableException.copy(UnavailableException.java:128)
	at com.datastax.driver.core.Responses$Error.asException(Responses.java:114)
	at com.datastax.driver.core.RequestHandler$SpeculativeExecution.onSet(RequestHandler.java:477)
	at com.datastax.driver.core.Connection$Dispatcher.channelRead0(Connection.java:1005)
	at com.datastax.driver.core.Connection$Dispatcher.channelRead0(Connection.java:928)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:339)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:324)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:254)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:339)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:324)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:339)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:324)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:242)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:339)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:324)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:847)
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:618)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:329)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:250)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:111)
	at java.lang.Thread.run(Thread.java:745)
Caused by: com.datastax.driver.core.exceptions.UnavailableException: Not enough replicas available for query at consistency LOCAL_ONE (1 required but only 0 alive)
	at com.datastax.driver.core.Responses$Error$1.decode(Responses.java:50)
	at com.datastax.driver.core.Responses$Error$1.decode(Responses.java:37)
	at com.datastax.driver.core.Message$ProtocolDecoder.decode(Message.java:266)
	at com.datastax.driver.core.Message$ProtocolDecoder.decode(Message.java:246)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:89)
	... 11 more
2016-06-03 12:28:57,473 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 1000 milliseconds
2016-06-03 12:28:57,487 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 1000 milliseconds
2016-06-03 12:29:02,939 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 2000 milliseconds
2016-06-03 12:29:02,970 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 2000 milliseconds
2016-06-03 12:29:12,945 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 4000 milliseconds
2016-06-03 12:29:12,974 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 4000 milliseconds
2016-06-03 12:29:17,947 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 8000 milliseconds
2016-06-03 12:29:17,977 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 8000 milliseconds
2016-06-03 12:29:28,481 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 16000 milliseconds
2016-06-03 12:29:28,974 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 16000 milliseconds
2016-06-03 12:29:44,482 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 32000 milliseconds
2016-06-03 12:29:44,975 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 32000 milliseconds
2016-06-03 12:30:16,482 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 64000 milliseconds
2016-06-03 12:30:16,975 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 64000 milliseconds
2016-06-03 12:31:20,483 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 128000 milliseconds
2016-06-03 12:31:20,976 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 128000 milliseconds
2016-06-03 12:33:28,484 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 256000 milliseconds
2016-06-03 12:33:28,976 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 256000 milliseconds
2016-06-03 12:37:44,484 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 512000 milliseconds
2016-06-03 12:37:44,977 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 512000 milliseconds
2016-06-03 12:46:16,485 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 600000 milliseconds
2016-06-03 12:46:16,977 ERROR com.datastax.driver.core.ControlConnection                    - [Control connection] Cannot connect to any host, scheduling retry in 600000 milliseconds
2016-06-03 12:46:54,906 INFO  org.apache.flink.runtime.taskmanager.Task                     - Attempting to cancel task Source: Custom Source (1/2)
2016-06-03 12:46:54,907 INFO  org.apache.flink.runtime.taskmanager.Task                     - Source: Custom Source (1/2) switched to CANCELING
2016-06-03 12:46:54,907 INFO  org.apache.flink.runtime.taskmanager.Task                     - Triggering cancellation of task code Source: Custom Source (1/2) (dec8d24e486ca9937739b7c6e07fbb05).
2016-06-03 12:46:54,909 INFO  org.apache.flink.runtime.taskmanager.Task                     - Attempting to cancel task Cassandra Sink (1/2)
2016-06-03 12:46:54,909 INFO  org.apache.flink.runtime.taskmanager.Task                     - Cassandra Sink (1/2) switched to CANCELING
2016-06-03 12:46:54,909 INFO  org.apache.flink.runtime.taskmanager.Task                     - Triggering cancellation of task code Cassandra Sink (1/2) (96511ef6293a893b0ef35dd211aea2b9).
2016-06-03 12:46:55,389 INFO  com.dataartisans.Job                                          - Received cancel in EventGenerator
2016-06-03 12:46:55,392 INFO  org.apache.flink.runtime.taskmanager.Task                     - Source: Custom Source (1/2) switched to CANCELED
2016-06-03 12:46:55,392 INFO  org.apache.flink.runtime.taskmanager.Task                     - Freeing task resources for Source: Custom Source (1/2)
2016-06-03 12:46:55,394 INFO  org.apache.flink.yarn.YarnTaskManager                         - Un-registering task and sending final execution state CANCELED to JobManager for task Source: Custom Source (dec8d24e486ca9937739b7c6e07fbb05)
2016-06-03 12:47:24,911 WARN  org.apache.flink.runtime.taskmanager.Task                     - Task 'Cassandra Sink (1/2)' did not react to cancelling signal, but is stuck in method:
 org.apache.flink.streaming.runtime.io.StreamInputProcessor.processInput(StreamInputProcessor.java:166)
org.apache.flink.streaming.runtime.tasks.OneInputStreamTask.run(OneInputStreamTask.java:65)
org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:224)
org.apache.flink.runtime.taskmanager.Task.run(Task.java:559)
java.lang.Thread.run(Thread.java:745)

2016-06-03 12:47:54,912 WARN  org.apache.flink.runtime.taskmanager.Task                     - Task 'Cassandra Sink (1/2)' did not react to cancelling signal, but is stuck in method:
 org.apache.flink.streaming.runtime.io.StreamInputProcessor.processInput(StreamInputProcessor.java:166)
org.apache.flink.streaming.runtime.tasks.OneInputStreamTask.run(OneInputStreamTask.java:65)
org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:224)
org.apache.flink.runtime.taskmanager.Task.run(Task.java:559)
java.lang.Thread.run(Thread.java:745)

2016-06-03 12:48:24,913 WARN  org.apache.flink.runtime.taskmanager.Task                     - Task 'Cassandra Sink (1/2)' did not react to cancelling signal, but is stuck in method:
 org.apache.flink.streaming.runtime.io.StreamInputProcessor.processInput(StreamInputProcessor.java:166)
org.apache.flink.streaming.runtime.tasks.OneInputStreamTask.run(OneInputStreamTask.java:65)
org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:224)
org.apache.flink.runtime.taskmanager.Task.run(Task.java:559)
java.lang.Thread.run(Thread.java:745)

2016-06-03 12:48:54,915 WARN  org.apache.flink.runtime.taskmanager.Task                     - Task 'Cassandra Sink (1/2)' did not react to cancelling signal, but is stuck in method:
 org.apache.flink.streaming.runtime.io.StreamInputProcessor.processInput(StreamInputProcessor.java:166)
org.apache.flink.streaming.runtime.tasks.OneInputStreamTask.run(OneInputStreamTask.java:65)
org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:224)
org.apache.flink.runtime.taskmanager.Task.run(Task.java:559)
java.lang.Thread.run(Thread.java:745)

2016-06-03 12:49:24,916 WARN  org.apache.flink.runtime.taskmanager.Task                     - Task 'Cassandra Sink (1/2)' did not react to cancelling signal, but is stuck in method:
 org.apache.flink.streaming.runtime.io.StreamInputProcessor.processInput(StreamInputProcessor.java:166)
org.apache.flink.streaming.runtime.tasks.OneInputStreamTask.run(OneInputStreamTask.java:65)
org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:224)
org.apache.flink.runtime.taskmanager.Task.run(Task.java:559)
java.lang.Thread.run(Thread.java:745)

2016-06-03 12:49:54,918 WARN  org.apache.flink.runtime.taskmanager.Task                     - Task 'Cassandra Sink (1/2)' did not react to cancelling signal, but is stuck in method:
 org.apache.flink.streaming.runtime.io.StreamInputProcessor.processInput(StreamInputProcessor.java:166)
org.apache.flink.streaming.runtime.tasks.OneInputStreamTask.run(OneInputStreamTask.java:65)
org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:224)
org.apache.flink.runtime.taskmanager.Task.run(Task.java:559)
java.lang.Thread.run(Thread.java:745)

2016-06-03 12:50:24,919 WARN  org.apache.flink.runtime.taskmanager.Task                     - Task 'Cassandra Sink (1/2)' did not react to cancelling signal, but is stuck in method:
 org.apache.flink.streaming.runtime.io.StreamInputProcessor.processInput(StreamInputProcessor.java:166)
org.apache.flink.streaming.runtime.tasks.OneInputStreamTask.run(OneInputStreamTask.java:65)
org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:224)
org.apache.flink.runtime.taskmanager.Task.run(Task.java:559)
java.lang.Thread.run(Thread.java:745)
```

theomega commented:
I tried out this branch and it works like it should in the scenario I set up:
I wrote (so only using the Sink) to a quiet complex columnfamily in a 8 node cassandra cluster. I was using a complex setup of windowed streams and all the data appeared and was perfectly readable as expected. I also had multiple sinks at the same time which also worked perfectly. I could not test the scenario @rmetzger is mentioning.

Overall, I agree with @rmetzger that it should be considered to merge this to get more users to test it and report issues.

rmetzger commented:
Thank you for testing the cassandra connector.

I'll merge the pull request now.

rmetzger commented:
The merging will probably be a bit delayed because there are some memory issues:

```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.apache.flink.streaming.connectors.cassandra.CassandraConnectorTest
06/13/2016 11:12:22	Job execution switched to status RUNNING.
06/13/2016 11:12:22	Source: Collection Source -> Sink: Unnamed(1/1) switched to SCHEDULED 
06/13/2016 11:12:22	Source: Collection Source -> Sink: Unnamed(1/1) switched to DEPLOYING 
06/13/2016 11:12:23	Source: Collection Source -> Sink: Unnamed(1/1) switched to RUNNING 
06/13/2016 11:12:27	Source: Collection Source -> Sink: Unnamed(1/1) switched to FINISHED 
06/13/2016 11:12:27	Job execution switched to status FINISHED.
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "CompactionExecutor:2"
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 150.02 sec - in org.apache.flink.streaming.connectors.cassandra.CassandraConnectorTest
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "main"
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "CompactionExecutor:5"
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "CompactionExecutor:4"
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "HintedHandoffManager:1"
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "CompactionExecutor:7"
==============================================================================
Maven produced no output for 300 seconds.
==============================================================================
==============================================================================
The following Java processes are running (JPS)
==============================================================================
2956 Launcher
98454 Jps
97148 surefirebooter7601064672285446339.jar
==============================================================================
Printing stack trace of Java process 2956
==============================================================================
2016-06-13 11:28:23
Full thread dump Java HotSpot(TM) 64-Bit Server VM (24.76-b04 mixed mode):
```

I will take care of them!

f1yegor commented:
Looking at the code of `CassandraSinkBase` I assume that every new sink will create own session to the cluster. 
I'm creating arount 1000 streams and sinks for them, so cassandra cluster would be saturated with openning connections.
Cassandra session could manage several connections via one session. 

Could you confirm that this isn't an issue with the implementation?

zentol commented:
You are correct that every CassandraSink instance opens a separate connection to the cluster. I don't see how we could avoid this without using Singletons, which should be avoided.

f1yegor commented:
Could somebody point to a documentation how we could manage state (pool of connections) when Sink works inside single JVM.
